Improve the REMOTE_USER (for example mod_auth_kerb) support

adelton opened the pull request:
This series of patches addresses http://projects.theforeman.org/issues/3312, making it possible to set up a protected "logon" location in Apache via

```
<Location /users/extlogin>
...
</Location>
```

and configure mod_auth_\* authentication modules for that location. It was tested with mod_auth_kerb.

The possibilities and issues with the existing REMOTE_USER support are listed at http://projects.theforeman.org/projects/foreman/wiki/Foreman_and_mod_auth_kerb, together with explanations introduced by this pull request.

This is a refactoring of https://github.com/theforeman/foreman/pull/958, addressing all comments in that pull request.

ares committed:
fixes #3312 - support /users/extlogin as a login page based on REMOTE_USER.

adelton committed:
fixes #3312 - when authorize_login_delegation_auth_source_user_autocreate is set, creates the unknown authenticated user.

adelton committed:
fixes #3312 - support logout from REMOTE_USER-based logons.

ares committed:
fixes #3312 - tweaking nil tests; private.

ares committed:
fixes #3312 - add tests.

dLobatog commented:
[test]

ares commented:
tests added in https://github.com/adelton/foreman/pull/2, worked on my machine, @adelton could you please merge/rebase?

adelton commented:
Rebased on develop, including Marek H.'s tests: 34defdc55765700eff0723e5852c2bc5ac1d2774.

lzap commented:
Thanks [test]

sodabrew commented:
Oh this is great, would it work with mod_auth_openid? https://github.com/bmuller/mod_auth_openid

adelton commented:
I did not verify it specifically with mod_auth_openid but the plan is to support anything that populates REMOTE_USER which mod_auth_openid does.

ares commented on app/assets/.directory:
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
```
oops, could you please remove .directory files? I accidentaly added them to my commit

adelton replied:
On Wed, Oct 23, 2013 at 11:05:57PM -0700, Marek Hulán wrote:

> > @@ -0,0 +1,4 @@
> > +[Dolphin]
> 
> oops, could you please remove .directory files? I accidentaly added them to my commit

Done, and rebased on latest develop: 1ecfb774ae6226e1bcf037e12d3b837a995beb85.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

adelton commented:
On Wed, Oct 23, 2013 at 11:05:57PM -0700, Marek Hulán wrote:

> > @@ -0,0 +1,4 @@
> > +[Dolphin]
> 
> oops, could you please remove .directory files? I accidentaly added them to my commit

Done, and rebased on latest develop: 1ecfb774ae6226e1bcf037e12d3b837a995beb85.

ares commented:
I went through this once again and when adelton/foreman/pull/3 gets in, you have my :+1: The only unresolved comment from #958 is about authorize_login_delegation_auth_source_user_autocreate type - @domcleal?

adelton commented:
The https://github.com/adelton/foreman/pull/3 is in e8426b38a6a05eb2fd3af9073ba618e5f6113631.

ohadlevy commented:
[test]

adelton commented:
What is preventing this pull request from being merged / rebased on develop?

ohadlevy commented on app/models/user.rb:
```diff
@@ -135,6 +135,25 @@ def self.try_to_login(login, password)
+        if auth_source.nil?
+          auth_source = AuthSourceExternal.new(:name => auth_source_name)
+          auth_source.save!
+        end
```
can be simplified:

``` ruby
options = { :name => auth_source_name }
auth_source = AuthSourceExternal.where(options).first || AuthSourceExternal.create!(options)
User.create!(:login => login, :auth_source => auth_source)
```

this first ensures we are always talking about External Authsources vs any authsource.

secondly, I'm wondering what happens if AutoSource is created, but user does not, I did not try, but we could try using the build method instead and save them in one go... maybe something like

``` ruby
options = { :name => auth_source_name }
auth_source = AuthSourceExternal.where(options).first || AuthSourceExternal.build(options)
User.create!(:login => login, :auth_source => auth_source)
```

AFAIK (require a bit testing) would create both user and auth source in the same transaction. (would nice to see a test like that).

adelton replied:
The goal is to be able to name any auth_source, not necessarily the external one.

If the build approach works, by all means go for it.

adelton replied:
Attempt to use the `build` approach leads to

```
 x undefined method `build' for #<Class:0x000000070bcb90> 
```

ohadlevy replied:
right.. i guess it should be auth_source.build(:type =>
'AuthSourceExternal'...)
but that would also need to include Foreman::STI to the AuthSource class
for it to work.

adelton replied:
Sorry, no idea what Foreman::STI is and how to include it in this patch. I'll go with the

```
options = { :name => auth_source_name }
auth_source = AuthSourceExternal.where(options).first || AuthSourceExternal.create!(options)
User.create!(:login => login, :auth_source => auth_source)
```

approach.

ohadlevy replied:
sure, but as you said, you might need to use AuthSource.where (vs.
AuthSourceExternal.where)

adelton replied:
Right. Pushed out as 12d820e2973891989068a13e683d5ce19d062267. Thanks.

ohadlevy commented on config/routes.rb:
```diff
@@ -183,6 +183,8 @@
         post 'login'
         get 'logout'
+        get 'extlogin'
+        get 'extlogout'
```
while not relevant to this PR.. I think its about time we refactor out *log{in/out} to a sessions controller.

ohadlevy commented on test/lib/foreman/access_permissions_test.rb:
```diff
@@ -12,7 +12,7 @@
 class AccessPermissionsTest < ActiveSupport::TestCase
   MAY_SKIP_REQUIRE_LOGIN = [
-    "users/login", "users/logout", "home/status", "notices/destroy", "unattended/",
+    "users/login", "users/logout", "users/extlogout", "home/status", "notices/destroy", "unattended/",
```
looks like these tests are failing?

adelton replied:
I have no idea what the situation with tests is. The change above was added based on Dominic's suggestion: https://github.com/theforeman/foreman/pull/958#issuecomment-26486270

domcleal replied:
You need to add users/extlogin as well now.

adelton replied:
Ah. Fixed in a22f033829e4167d7f76a7e49141eb6220fc1ff5. Thanks.

adelton committed:
fixes #3312 - factor out the post-login logic from try_to_login so that it can be used for external logins as well.

adelton commented:
Rebased on latest develop and added commit 8a1f17c9fa03d51f9a28fd94362a37d372975f4a to give the new user the Anonymous role upon login, like form-based logons do.

adelton commented:
New rebase to address Ohad's comments: 59f9484a2ed83a9b52509d1e181c46afd9ed61ad.

domcleal commented:
I'm having problems testing this after seeing up Foreman as per the wiki page.  My browser begins looping around the /users/extlogin page with a series of 302 and 401 HTTP responses.

http://people.redhat.com/~dcleal/extlogin_redirect.png (browser logs)
http://pastie.org/private/zaponz70lg0pdszlcmfoa (httpd logs)
http://pastie.org/private/ngudp6v82kjyljrrraxnng (production.log)

Any thoughts?

ares commented:
Production log contains only 302 so 401 is rendered by apache before it gets to foreman? How did you set the `Setting['login_delegation_logout_url']`? Did you set kerberos login only on `/users/extlogin` sub-uri?

domcleal commented:
I hadn't set the logout_url setting, but have now have - it makes no difference:

```
:login_delegation_logout_url: https://rhel6c.ipa.fm.example.net/users/extlogout
```

Yes, I'm only using mod_auth_kerb on /users/extlogin using /etc/httpd/conf.d/auth_kerb.conf:

<pre>
 &lt;Location /users/extlogin&gt;
 AuthType Kerberos
 AuthName "Kerberos Login" 
 KrbMethodNegotiate On
 KrbMethodK5Passwd Off
 KrbAuthRealms IPA.FM.EXAMPLE.NET
 Krb5KeyTab /etc/http.keytab
 KrbLocalUserMapping On
 require valid-user
 ErrorDocument 401 '<html><meta http-equiv="refresh" content="0; URL=/users/login"><body>Kerberos authentication did not pass.</body></html>'
 ErrorDocument 500 '<html><meta http-equiv="refresh" content="0; URL=/users/login"><body>Kerberos authentication did not pass.</body></html>' 
 &lt;/Location&gt;
</pre>

adelton commented:
My output after successful Kergeros login is

```
Started GET "/users/extlogin" for 10.36.112.30 at 2013-11-04 20:54:46 -0500
Processing by UsersController#extlogin as HTML
Authorized user admin(Admin User)
Redirected to https://foreman.example.com/hosts
Completed 302 Found in 21ms (ActiveRecord: 11.7ms)
```

Since I don't see that `Authorized user ...` message in http://pastie.org/private/ngudp6v82kjyljrrraxnng I assume something has gone wrong in your setup. In any case, I am not able to reproduce that loop of yours (I just did another fresh installation from http://yum.theforeman.org/releases/1.3/el6/$basearch).

adelton commented:
I assume the culprit could be in

```
  def login_user(user)
    session[:user]         = user.id
    uri                    = session[:original_uri]
    session[:original_uri] = nil
    redirect_to (uri || hosts_path)
  end
```

-- if you have `/users/extlogin` in `session[:original_uri]`, that might cause the loop. Can you tweak the code to check if that is the case?

But I have no idea how that `/users/extlogin` would get there in the first place.

adelton commented:
Did you have chance to test this? If the endless loop comes from this code, we might want to not do the `redirect_to` if the `uri` is already the same as `request.fullpath`.

domcleal commented:
PEBKAC, sorry for the noise.  I'd done a nightly install and attempted to apply the patch on top, but somehow it hadn't fully applied.  An RPM build of the branch worked much better.

domcleal commented:
Rebased and merged as b7589c3 for Foreman 1.4.0.  Thanks very much @adelton, and my apologies for the delay.

sodabrew commented:
Awesome work! I'm very excited about this merge, thanks @adelton!

adelton commented:
Thanks for the merge. I'll now rebase https://github.com/theforeman/foreman/pull/986 and work on it.
